### DISCUSSION TITLE
Executing Astro component scripts

### DISCUSSION THREAD

**Discussion by kooalab** [swup/swup#856] [Help] [answered]
### Describe your problem 🧐

I'm integrating Astro with swup.js in my project, where each component contains script tags. 

The issue arises during page transitions; the JavaScript within these components does not execute. 

To address this, I implemented swup.js's Head Plugin, anticipating it would reload necessary JavaScript, as in Astro, JavaScript is typically loaded from the head. Despite this, the expected functionality isn't achieved. 

I've created a demo on StackBlitz to illustrate the problem. 

### What is your question? 🙋‍♀️

How can I ensure that JavaScript within each component executes correctly during page transitions in an Astro project integrated with swup.js, especially when using the Head Plugin? 

Despite implementing this plugin, which I thought would reload the necessary JavaScript from the head section during page transitions, the scripts in the components are not executing. 

Is there a specific configuration or approach I should use to enable seamless JavaScript execution across page transitions in this setup?

Thanks in advance for your help 🙏🏼 Really appreciated...

### URL to minimal reproduction 🔗

https://stackblitz.com/edit/withastro-astro-3jsbn4?file=src%2Fpages%2Findex.astro

### Checked all these? 📚

- [X] I have read and searched the [official docs](https://swup.js.org)
- [X] I have checked [discussions](https://github.com/orgs/swup/discussions) and [existing issues](https://github.com/issues?q=is:issue+org:swup) for answers to my question
- [X] I agree to follow this project's [Code of Conduct](https://github.com/swup/.github/blob/master/CODE_OF_CONDUCT.md)
- [X] I have provided all necessary information to the best of my knowledge

**Comment by daun** [comment, marked as answer]
To insert new scripts and run them as well, you can combine [Head Plugin](https://github.com/swup/head-plugin) and [Scripts Plugin](https://github.com/swup/scripts-plugin). That's how the [official Astro integration](https://github.com/swup/astro) does it, and it's working great. Quoting the docs on [Common Issues](https://swup.js.org/getting-started/common-issues/):

> Use the **Head Plugin** to insert new scripts into the head  
> Use the **Scripts Plugin** to run existing and new scripts in head or body

**Reply by kooalab** [reply]
Hey @daun,

Thank you for the quick response. 

I've already tried using both the Head Plugin and Scripts Plugin, as recommended, but I'm still facing issues. 

I've updated my [StackBlitz demo ](https://stackblitz.com/edit/withastro-astro-3jsbn4?file=src%2Flayouts%2FLayout.astro)to include the Scripts Plugin setup. Unfortunately, I'm still struggling with getting the JavaScript to execute correctly during page transitions. I'm reaching a point of desperation, as I've tried everything I can think of, and it's still not working. 

Could you please take a look at the updated demo and see if there's anything immediately apparent that I'm missing or doing wrong? Your help would be greatly appreciated, as I'm at a loss on how to proceed.

**Reply by daun** [reply]
In your component, you're listening for `DOMContentLoaded`, which will only ever fire once. On subsequent executions, the alert won't show as that event is not triggered again unless you manually reload the tab. That's the case with all frontend routers that avoid full page reloads. There's a section on the docs with an explanation and some examples: [Reloading JavaScript](https://swup.js.org/getting-started/reloading-javascript/).

**Reply by daun** [reply]
On that page, there's example code for handling that. It's all in the docs :)

```js
// Run once when page loads
if (document.readyState === 'complete') {
  init();
} else {
  document.addEventListener('DOMContentLoaded', () => init());
}
```

**Reply by kooalab** [reply]
Thanks again @daun. My Astro project is set up to load JavaScript per component rather than in one large bundle. Each component includes its own script tags relevant to its functionality, taking advantage of Astro's on-demand script loading for the current page. 
The challenge I'm encountering is figuring out how to reinitialize these scripts with swup after a page transition. Since the scripts are not centralized, I'm unsure how to apply the init pattern across multiple, separate components. 
Any advice on managing this with the swup instance initialized in layout.astro would be immensely helpful. Should I use something like nanostores to share the swup instance across all of my components so that I have access to that "swup.hooks.on('page:view'"?

Sorry for these noob questions, I'm relatively new to this 😅

**Reply by daun** [reply]
If you use that pattern in each of your components, you should be fine. This is unrelated to whether you load one large script or many smaller ones, it will work in both cases. If you need access to swup's hooks, you could store the instance on the window object: `window.swup = new Swup()`. That's not super clean, but it works.

**Reply by kooalab** [reply]
I've tried the window.swup but I've hit a bit of a snag, though. There's a timing issue where window.swup is sometimes undefined in components that execute before it's been set up in layout.astro. It's a classic race condition. I’m going to keep at it and try to find a workaround that maintains the neatness of each component handling its own JavaScript alongside swup. I'll loop back here with any breakthroughs. Meanwhile, if a spark of inspiration hits you or if you have another trick up your sleeve, I’m all ears! Thanks again for your help 🙏🏼

**Reply by daun** [reply]
Swup triggers [dom events](https://swup.js.org/hooks/#dom-events) as well, you can listen for those:

```js
if (window.swup) {
  // swup is available right away
} else {
  document.addEventListener('swup:enable', () => {
    // swup is available now
  });
}

**Reply by daun** [reply]
Since you're mentioning neatness — you can avoid all that mess by using a component framework that self-initializes, like Alpine.js, Stimulus, or Gia. See [Using component frameworks](https://swup.js.org/getting-started/reloading-javascript/#using-component-frameworks) in the docs. It's really the simplest and cleanest approach to managing dynamically loaded components.

**Reply by kooalab** [reply]
Yes this is a mess... I should definitely look into alpine.js or qwick to make this the proper way.

Just to be sure, if I have this
` 
if (window.mySwup) {
    window.mySwup.hooks.on("page:view", init);
  } else {
    document.addEventListener("swup:enable", () => {
      window.mySwup.hooks.on("page:view", init);
    });
  }
`
For every component that has JS that needs to be reexecuted on each page visit, I won't run in performance issues?

**Reply by daun** [reply]
Depends on what you're doing inside `init`, but the event boilerplate shouldn't be a performance bottleneck.

**Reply by kooalab** [reply]
I just realized that what I had in mind can't actually work... My goal was to not have the JS in one big file but rather splitting it in every component where it is needed. 

But doing it like that, if the user starts on the index.astro page, its going to load some JS that is needed for that page, but then, if I go let's say to the about page that has his own components with their own JS, swup will replace the main container but the JS of this new components that weren't present in the inital JS file that have this : 
`
 if (window.mySwup) {
     window.mySwup.hooks.on("page:view", init);
  } else {
    document.addEventListener("swup:enable", () => {
      window.mySwup.hooks.on("page:view", init);
    });
  }
`
will actually never get executed because the code will never run?

**Reply by kooalab** [reply]
Oh that's where the Scripts plugin comes in. Got it, thank you 😅
